libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/Store.kt

```diff
@@ -32,6 +32,19 @@ fun interface Reducer {
     fun process(state: State, action: Action): State
 }
 
+/**
+ * Similar to the [Action], except these actions complete asynchronous tasks such as network
+ * requests within a coroutine scope.
+ */
+interface SuspendAction
+
+/**
+ * Similar to the [Reducer], except the state is processed and completed with a result.
+ */
+fun interface SuspendReducer {
+    suspend fun process(state: State, action: SuspendAction): State
```

Comment:
This is an interesting solution. However, it might easily lead to a store state corruption. 

Imagine the following situation:
Given Reducer A, Reducer B and actions A1, A2.

![Async Reducer](https://user-images.githubusercontent.com/2678039/182401992-7d67e683-3bc9-4c4e-b0ce-c0bfaa7b53b6.png)

1. **UIComponent** dispatches an action (**A1**).
2. **Store** forwards action (**A1**) + state (**S1**) to a Suspend **Reducer A** for processing.
3. **Reducer A** captures given state (**S1**) and starts async operation.
4. Meanwhile, **UIComponent** dispatches another action (**A2**).
5. **Store** forwards action (**A2**) + state (**S1**) to another **Reducer B**
6. **Reducer B** synchronously processes the action and returns a new state (_S1 -> S2_).
7.  The **Asynch Operation** finishes and **Reducer A** mutates previously captured state (_S1 -> S1'_).

The information stored in state S2 is lost and the state ends up in a corrupted state (S1').

Reply:
The diagram shows typical race condition with asynchronous workers. The solution is typically a locking mechanism, which we could add but often it is unnecessary. At least with suspending actions, the caller is able to block until the action is complete. Without suspending actions, the caller does not have a way to detect when an action completes.

This same issue would exist on with `ThunkAction` right? Or are you [saying this approach handles](https://github.com/mapbox/mapbox-navigation-android/pull/6069/files) this case?

Reply:
[This approach](https://github.com/mapbox/mapbox-navigation-android/pull/6069/files#diff-6beb6c390c9dcbc0f82ce3c11decc73f8e83b04a8e909f5960554f637b94cbafR111) handles it by passing an **Async Operation** result via Action instead of state change.

The critical thing to remember with Flux architecture is that only Reducers are allowed to mutate the state, and this change always has to be done synchronously to avoid information loss.

Reply:
But nothing is preventing two **Async Operations** to attempt to alter the same state and creating a race condition.

For example, here is a race condition from the `ThunkAction`
``` kotlin
binding.routePreview.onClick(coroutineScope) {
  store.dispatch(showRoutePreview())
  store.dispatch(startActiveNavigation())
}
```

The caller has no way to determine when the `showRoutePreview` call has completed. You would think that this would `startActiveNavigation`, but it almost always shows the route preview (unless the route has been fetched)

Reply:
With the approach I'm proposing, the **caller** has control of when the suspending action completes

So, you remove the race condition. Below will go into ActiveNavigation as expected

```
coroutineScope.launch {
    if (store.process(RoutesSuspendAction.RequestCurrent).routes is RoutesState.Ready) {
        store.dispatch(NavigationStateAction.Update(NavigationState.RoutePreview))
        store.dispatch(NavigationStateAction.Update(NavigationState.ActiveNavigation))
    }
}
```

Reply:
> The caller has no way to determine when the `showRoutePreview` call has completed. You would think that this would `startActiveNavigation`, but it almost always shows the route preview (unless the route has been fetched)

```kotlin
binding.routePreview.onClick(coroutineScope) {
  store.dispatch(showRoutePreview())
  store.dispatch(startActiveNavigation())
}
```

Reply:
> But nothing is preventing two **Async Operations** to attempt to alter the same state and creating a race condition.
> 
> For example, here is a race condition from the `ThunkAction`
> 
> ```kotlin
> binding.routePreview.onClick(coroutineScope) {
>   store.dispatch(showRoutePreview())
>   store.dispatch(startActiveNavigation())
> }
> ```
> 
> The caller has no way to determine when the `showRoutePreview` call has completed. You would think that this would `startActiveNavigation`, but it almost always shows the route preview (unless the route has been fetched)

In flux architecture, "the caller" (_UIComponent_) has only two responsibilities: 
1. updating UI when **State** changes
2. dispatch appropriate **Actions** as a response to user Input (_onClick_)

It is not responsible for control flow. That's the controller's (_StateController_) responsibility.

However, if for some reason we want "the caller" to control the flow (without rewriting both thunks), the above code needs to be slightly amended. We need ensure the system is in the right state before dispatching `startActiveNavigation())`.

```kotlin
binding.routePreview.onClick {
  scope.launch {
      store.dispatch(showRoutePreview())
      store.waitWhileFetching()
      store.dispatch(startActiveNavigation())
  }
}
...
private suspend fun Store.waitWhileFetching() {
    select { it.previewRoutes }.takeWhile { it is RoutePreviewState.Fetching }.collect()
}
```

A better approach would be to replace both thunks with actions of the same name, and move their processing logic to the controller (similar to [fetchRouteSaga](https://github.com/mapbox/mapbox-navigation-android/pull/6069/files#diff-6df081b485d4b11864caa00f184174bf37c271dabf3bdab7ea2a607845faf6a8R85)). 

> With the approach in this pull request, the **caller** has control of when the suspending action completes
> 
> So, you remove the race condition. Below will go into ActiveNavigation as expected
> 
> ```
> coroutineScope.launch {
>     if (store.process(RoutesSuspendAction.RequestCurrent).routes is RoutesState.Ready) {
>         store.dispatch(NavigationStateAction.Update(NavigationState.RoutePreview))
>         store.dispatch(NavigationStateAction.Update(NavigationState.ActiveNavigation))
>     }
> }
> ```

The approach proposed in this PR violates Flux Architecture. In Flux: 
- dispatch is a fire-and-forget, non-blocking operation
- state can only be changed by a Reducer
- view can only be updated as a result of state change

To better explain why it's designed like this, let me quote a description from the ["Flux In-Depth Overview" page](https://facebook.github.io/flux/docs/in-depth-overview/)

> ![](https://facebook.github.io/flux/img/overview/flux-simple-f8-diagram-explained-1300w.png)
> (...)
> _This structure allows us to reason easily about our application in a way that is reminiscent of functional reactive programming, or more specifically data-flow programming or flow-based programming, where data flows through the application in a single direction — there are no two-way bindings. Application state is maintained only in the stores, allowing the different parts of the application to remain highly decoupled._
> (...)

Reply:
>>  binding.routePreview.onClick(coroutineScope) {
  store.dispatch(showRoutePreview())
  store.dispatch(startActiveNavigation())
}
The caller has no way to determine when the showRoutePreview call has completed. You would think that this would startActiveNavigation, but it almost always shows the route preview (unless the route has been fetched)

> In flux architecture, "the caller" (UIComponent) has only two responsibilities:

🤔 .. i thought this was a pretty good point where, when the caller calls these two simple functions the expected state would be the second one. Why is the caller so concerned with flux architecture?

Reply:
```
private suspend fun Store.waitWhileFetching() {
    select { it.previewRoutes }.takeWhile { it is RoutePreviewState.Fetching }.collect()
}
```

Ok concept am looking for with suspending actions, was to make it so these type of functions are supported api. Rather than private extensions, so maybe we can meet in the middle somewhere?

Reply:
Not making much progress in this conversation so am closing this PR and letting you run with it @tomaszrybakiewicz